File: CS/WindowsApplication1/XtraReport1.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using DevExpress.XtraReports.UI;
using System.Data;
using System.Collections.Generic;
using System.Linq;

namespace WindowsApplication1 {
    public partial class XtraReport1: DevExpress.XtraReports.UI.XtraReport {
        public XtraReport1() {
            InitializeComponent();
        }

        private void calculatedField1_GetValue(object sender,GetValueEventArgs e) {
            DetailItemClass currentRow = e.Row as DetailItemClass;
            List<ParentItemClass> ds = (List<ParentItemClass>)e.Report.DataSource;
            List<DetailItemClass> childRows = ds.Single(x => x.ItemID == currentRow.ParentID).DetailItems;
            e.Value = GetCurrentRowIndex(currentRow,childRows);
        }

        private int GetCurrentRowIndex(DetailItemClass row, List<DetailItemClass> rows) {
            int currentTimeStampIndex = 1;
            int numberOfFullRows = rows.Count / 3;
            int numberofAdditionalCells = rows.Count % 3;
            for(int i = 0; i < rows.Count; i++) {
                if(((TimeSpan)rows[i].TimeStamp) < row.TimeStamp) {
                    currentTimeStampIndex++;
                }
            }
            int columnIndex = 1;
            int cellsCount = 0;
            for(columnIndex = 1; columnIndex < 4; columnIndex++) {
                int currentColumnCellsCount = 0;
                if(columnIndex > numberofAdditionalCells) {
                    cellsCount += numberOfFullRows;
                    currentColumnCellsCount = numberOfFullRows;
                } else {
                    cellsCount += numberOfFullRows + 1;
                    currentColumnCellsCount = numberOfFullRows + 1;
                }
                if(currentTimeStampIndex <= cellsCount) {
                    int rowindex = currentTimeStampIndex - (cellsCount - currentColumnCellsCount);
                    return (rowindex - 1) * 3 + columnIndex;
                }
            }
            return -1;
        }

    }
}

File: VB/WindowsApplication1/XtraReport1.vb
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports DevExpress.XtraReports.UI
Imports System.Data
Imports System.Collections.Generic
Imports System.Linq

Namespace WindowsApplication1
    Partial Public Class XtraReport1
        Inherits DevExpress.XtraReports.UI.XtraReport

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub calculatedField1_GetValue(ByVal sender As Object, ByVal e As GetValueEventArgs)
            Dim currentRow As DetailItemClass = TryCast(e.Row, DetailItemClass)
            Dim ds As List(Of ParentItemClass) = CType(e.Report.DataSource, List(Of ParentItemClass))
            Dim childRows As List(Of DetailItemClass) = ds.Single(Function(x) x.ItemID = currentRow.ParentID).DetailItems
            e.Value = GetCurrentRowIndex(currentRow, childRows)
        End Sub

        Private Function GetCurrentRowIndex(ByVal row As DetailItemClass, ByVal rows As List(Of DetailItemClass)) As Integer
            Dim currentTimeStampIndex As Integer = 1
            Dim numberOfFullRows As Integer = rows.Count / 3
            Dim numberofAdditionalCells As Integer = rows.Count Mod 3

            For i As Integer = 0 To rows.Count - 1

                If (CType(rows(i).TimeStamp, TimeSpan)) < row.TimeStamp Then
                    currentTimeStampIndex += 1
                End If
            Next

            Dim columnIndex As Integer = 1
            Dim cellsCount As Integer = 0

            For columnIndex = 1 To 4 - 1
                Dim currentColumnCellsCount As Integer = 0

                If columnIndex > numberofAdditionalCells Then
                    cellsCount += numberOfFullRows
                    currentColumnCellsCount = numberOfFullRows
                Else
                    cellsCount += numberOfFullRows + 1
                    currentColumnCellsCount = numberOfFullRows + 1
                End If

                If currentTimeStampIndex <= cellsCount Then
                    Dim rowindex As Integer = currentTimeStampIndex - (cellsCount - currentColumnCellsCount)
                    Return (rowindex - 1) * 3 + columnIndex
                End If
            Next

            Return -1
        End Function
    End Class
End Namespace
